File: cpp/cpp-pthread/demo11b-exec-service.cpp
/*
EXECUTOR SERVICES AND THREAD POOLS

Executor services in C++ POSIX threading are not supported by default.
So, I use mylib::ExecService for this demonstration.
*/


#include <iostream>
#include <unistd.h>
#include "mylib-execservice.hpp"
using namespace std;



int main() {
    constexpr int NUM_THREADS = 2;
    constexpr int NUM_TASKS = 5;

    auto execService = mylib::ExecService(NUM_THREADS);

    for (int i = 0; i < NUM_TASKS; ++i) {
        execService.submit([=] {
            char id = 'A' + i;
            cout << "Task " << id << " is starting" << endl;
            sleep(3);
            cout << "Task " << id << " is completed" << endl;
        });
    }

    cout << "All tasks are submitted" << endl;

    execService.waitTaskDone();
    cout << "All tasks are completed" << endl;

    execService.shutdown();

    return 0;
}

File: js-nodejs/demo11b-exec-service.js
/*
EXECUTOR SERVICES AND THREAD POOLS
*/

import * as mylib from './mylib.js';
import { isMainThread } from 'worker_threads';
import { Piscina } from 'piscina';


//---------------------------------------------
//            WORKER THREAD SECTION
//---------------------------------------------


export const doTask = async ({index}) => {
  const id = String.fromCharCode(65 + index);
  console.log(`Task ${id} is starting`);
  await mylib.sleep(3000);
  console.log(`Task ${id} is completed`);
};


//---------------------------------------------
//             MAIN THREAD SECTION
//---------------------------------------------


const mainFunc = async () => {
  try {
    const NUM_THREADS = 2;
    const NUM_TASKS = 5;
    const lstPromRes = [];

    const execService = new Piscina({
      filename: new URL(import.meta.url).href,
      minThreads: NUM_THREADS,
      maxThreads: NUM_THREADS
    });

    for (let i = 0; i < NUM_TASKS; ++i) {
      lstPromRes.push(execService.run({ index: i }, { name: 'doTask'}));
    }
    console.log('All tasks are submitted');

    await Promise.all(lstPromRes);
    console.log('All tasks are completed');

    await execService.destroy();

  } catch (error) {
    console.error(error);
  }
};


if (isMainThread) {
  await mainFunc();
}
